### ISSUE TITLE
For @NewSpan annotation, Flux types are generating a new trace rather than a child span.

### ISSUE THREAD

**Issue by adirepo** [spring-cloud/spring-cloud-sleuth#1634]
When I add _@NewSpan_ annotation on methods returning Flux, it generates a new trace all together rather than a new child span under the current trace.

![Capture](https://user-images.githubusercontent.com/30016801/82026244-c99b2100-9692-11ea-8ab4-2af8f3795911.JPG)

**Comment by adriancole** [comment]
as this is very specific to how you use newspan, can you make a failing unit test here?

Otherwise, fork below, demonstrating the break in simplest form? Ex don't change anything except using newspan in a way that breaks.
https://github.com/openzipkin/sleuth-webmvc-example#spring-webflux-tracing

If neither, can you paste what you are doing?

**Comment by spring-issuemaster** [comment]
If you would like us to look at this issue, please provide the requested information. If the information is not provided within the next 7 days this issue will be closed.

**Comment by jstanik** [comment]
I experienced a similar issue when a new trace is generated instead of spawning a child span. But the method does not have to even return a reactive type. It's sufficient to use the operator `subscribeOn(Schedulers.boundedElastic)` on a reactive type. Here is a short example:

```java
@Component
class Foo {

  @Autowired
  Bar bar;

  public String doFoo() {
    return Mono.fromCallable(bar::doBar)
        .subscribeOn(Schedulers.boundedElastic())
        .block();
  }
}

@Component
class Bar {

  @NewSpan
  public String doBar() {
    return "bar";
  }
}

```

I identified the problem being at line 55 of  [SleuthContextListener](https://github.com/spring-cloud/spring-cloud-sleuth/blob/c80e1450652d285a1dd70dbe6c03a6d246ce64ad/spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/async/SleuthContextListener.java#L55) which creates new instance of `SleuthContextListener` even when it should return an existing instance. The new instance always reports the context hasn't been refreshed. The method will not find the existing instance in the `CACHE` because it was put to the cache on line 75 as
```java
if (context instanceof ConfigurableApplicationContext) {
  beanFactory = ((ConfigurableApplicationContext) context).getBeanFactory();
}
``` 
The method `getBean(BeanFactory)` should also look like:

```java
static SleuthContextListener getBean(BeanFactory beanFactory) {
  if (beanFactory instanceof ConfigurableApplicationContext) {
    beanFactory = ((ConfigurableApplicationContext) beanFactory).getBeanFactory();
  }
  return CACHE.getOrDefault(beanFactory, new SleuthContextListener());
}
```

This incorrect behavior causes [ContextUtil.isContextUnusable(BeanFactory)](https://github.com/spring-cloud/spring-cloud-sleuth/blob/c80e1450652d285a1dd70dbe6c03a6d246ce64ad/spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/async/ContextUtil.java#L38) to report the context is unusable even though the context was refreshed and is not closed which changes behavior in many places. For example the [TraceableScheduledExecutorService.schedule](https://github.com/spring-cloud/spring-cloud-sleuth/blob/c80e1450652d285a1dd70dbe6c03a6d246ce64ad/spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/async/TraceableScheduledExecutorService.java#L49) method at line 49 where it instead of scheduling the `TraceRunnable` schedules the raw `Runnable` which causes a new trace is created instead of creating a child span.

**Comment by adirepo** [comment]
I am not sure if this information would help, I think the problem might be with `ReactorSleuthMethodInvocationProcessor` class. I think **subscribe** method in **FluxSpan** is creating a new trace on line 149 

`span = tracer.newTrace();`

However, for **MonoSpan** on line 196 it is `span = tracer.nextSpan();`

In previous version of sleuth core 2.2.1 it was `span = tracer.nextSpan()` in **FluxSpan** class. 

My code used to give me desired result of child span creation for Flux types, but with current version sleuth it generates a new trace.

**Comment by adriancole** [comment]
here's the fix good eye.. we still need a relevant failing test.

**Comment by adriancole** [comment]
https://github.com/spring-cloud/spring-cloud-sleuth/pull/1656

**Comment by adriancole** [comment]
@jstanik please re-open a new issue for your comment as it is unrelated to this even if it is a bug. Easier to track different problems in different issues.

**Comment by jstanik** [comment]
@adriancole thanks for giving me the right direction. I was not sure if it was a different bug because the symptoms were the same.

I will open a new issue.

**Comment by adriancole** [comment]
@jstanik yeah I'm not sure it is a bug either, but if it is one, it isn't this one :D

**Comment by adriancole** [comment]
done and merged also to master (3.0.x) thanks for the help figuring this out @adirepo. Made mince meat out of the problem, knowing where to look!

**Comment by adirepo** [comment]
You are welcome @adriancole :). 
Thanks for your swift response and closure.
